robot-server/robot_server/hardware.py

```diff
@@ -0,0 +1,152 @@
+"""Hardware API initialization and management."""
+import asyncio
+import logging
+from pathlib import Path
+from fastapi import Depends, status
+from typing import cast
+from typing_extensions import Literal
+
+from opentrons import ThreadManager, initialize as initialize_api
+from opentrons.config import feature_flags
+from opentrons.util.helpers import utc_now
+from opentrons.hardware_control import API as HardwareAPI
+from opentrons.hardware_control.simulator_setup import load_simulator
+from opentrons.hardware_control.types import HardwareEvent, HardwareEventType
+
+from notify_server.clients import publisher
+from notify_server.settings import Settings as NotifyServerSettings
+from notify_server.models import event, topics
+from notify_server.models.hardware_event import DoorStatePayload
+
+from .app_state import AppState, get_app_state
+from .errors import ErrorDetails
+from .settings import get_settings
+
+_HARDWARE_API_KEY = "hardware_api"
+_HARDWARE_API_INIT_TASK_KEY = "hardware_api_init_task"
+_HARDWARE_EVENT_UNSUBSCRIBE_KEY = "hardware_event_unsubscribe"
+
+log = logging.getLogger(__name__)
+
+
+class HardwareNotYetInitialized(ErrorDetails):
+    """An error when accessing the HardwareAPI before it's initialized."""
+
+    id: Literal["HardwareNotYetInitialized"] = "HardwareNotYetInitialized"
+    title: str = "Hardware Not Yet Initialized"
+    detail: str = "The device's hardware has not finished initializing."
+
+
+class HardwareFailedToInitialize(ErrorDetails):
+    """An error if the HardwareAPI fails to initialize."""
+
+    id: Literal["HardwareFailedToInitialize"] = "HardwareFailedToInitialize"
+    title: str = "Hardware Failed to Initialize"
+
+
+def initialize_hardware(app_state: AppState) -> None:
+    """Initialize the HardwareAPI singleton, attaching it to global state."""
+    initialize_task = getattr(app_state, _HARDWARE_API_INIT_TASK_KEY, None)
+
+    if initialize_task is None:
+        initialize_task = asyncio.create_task(_initialize_hardware_api(app_state))
+        setattr(app_state, _HARDWARE_API_INIT_TASK_KEY, initialize_task)
+
+
+async def cleanup_hardware(app_state: AppState) -> None:
+    """Shutdown the HardwareAPI singleton and remove it from global state."""
+    initialize_task = getattr(app_state, _HARDWARE_API_INIT_TASK_KEY, None)
+    hardware_api = getattr(app_state, _HARDWARE_API_KEY, None)
+    unsubscribe_from_events = getattr(app_state, _HARDWARE_EVENT_UNSUBSCRIBE_KEY, None)
+
+    setattr(app_state, _HARDWARE_API_INIT_TASK_KEY, None)
+    setattr(app_state, _HARDWARE_API_KEY, None)
+    setattr(app_state, _HARDWARE_EVENT_UNSUBSCRIBE_KEY, None)
+
+    if initialize_task is not None:
+        initialize_task.cancel()
+        await asyncio.gather(initialize_task, return_exceptions=True)
```

Comment:
I think this can just be this?

```suggestion
        await initialize_task
```

Reply:
This is to ensure shutdown happens cleanly if:

- Initialization raised for some reason
- Shutdown requested before the hardware initialization finishes

Reply:
Aha. I think I'd find a `try`/`except` clearer, but this works for me if you want to keep it as-is.